crates/ty_server/src/server/api/semantic_tokens.rs

```diff
@@ -0,0 +1,61 @@
+use lsp_types::SemanticToken;
+use ruff_db::source::{line_index, source_text};
+use ruff_text_size::{Ranged, TextLen, TextRange};
+use ty_ide::semantic_tokens;
+use ty_project::ProjectDatabase;
+
+use crate::document::{PositionEncoding, TextSizeExt};
+
+/// Common logic for generating semantic tokens, either for full document or a specific range.
+/// If no range is provided, the entire file is processed.
+pub(crate) fn generate_semantic_tokens(
+    db: &ProjectDatabase,
+    file: ruff_db::files::File,
+    range: Option<TextRange>,
+    encoding: PositionEncoding,
+) -> Vec<SemanticToken> {
+    let source = source_text(db, file);
+    let line_index = line_index(db, file);
+
+    let requested_range = range.unwrap_or_else(|| TextRange::new(0.into(), source.text_len()));
+    let semantic_token_data = semantic_tokens(db, file, Some(requested_range));
```

Comment:
I would prefer passing `None` for the `requested_range` instead of a range covering the full document. It's more what I would expect, given that the `semantic_tokens` API already takes an `Option`